### DISCUSSION TITLE
GraphQL ability to set transform immediately false by default

### DISCUSSION THREAD

**Discussion by Wiejeben** [craftcms/cms#8895] [Ideas]
### Description
Title says all. Oddly enough `generateTransformsBeforePageLoad` is overwritten by `\craft\controllers\GraphqlController::actionApi`.

I don't want to accidentally have it be immediately transformed, especially on queries that result with 20+ images. I can already see someone accidentally forgetting to provide the `immediately: false` param in the future on the projects I work on.

**Comment by brandonkelly** [comment]
> Oddly enough `generateTransformsBeforePageLoad` is overwritten by `\craft\controllers\GraphqlController::actionApi`.

To clarify, this is expected because that setting is only meant to apply to web pages. API requests are more likely to be cached by the client, and we wouldn’t want to risk the transform generation URL getting cached (unless you’ve explicitly passed `immediately: false` in the query).

**Comment by Wiejeben** [comment]
Didn't think about the caching point of view. Actually pretty good reminder to myself when using cached GraphQL responses (in the frontend layer)!

**Comment by timkelty** [comment]
@brandonkelly this can probably be combined with https://github.com/craftcms/cms/issues/6387

**Comment by brandonkelly** [comment]
@timkelty Ah yes. I’ll close yours in favor of this one though, since this one’s a bit more apt.

**Comment by Wiejeben** [comment]
Oops didn't see that one. Having a `generateTransformsBeforeGqlResponse` config would be the perfect solution to me!

**Comment by Wiejeben** [comment]
I'd love to have this as a global setting so on my slower local setup I would be able to get transform URLs but on the more performant production server I would always get back a file URL.

File URLs can be stored in browser cache but the transform URL (which will stay with cached GraphQL responses) makes it unable to leverage browser caching.

**Comment by kara-todd** [comment]
I'm still running into the issue brought up in #5553 . In my case, there are a few large transforms that need to be done, and I'm using GraphQL to live reload data in a preview. I don't want to wait on the transforms ahead of time as they block the rest of the preview data from being viewed. I'd rather just load it with the transform url and wait for the request (similar to viewing on the front-end)

However, the transforms are coming back with my site url instead of my cp url and so the transforms never resolve. Is there a workaround for this?

**Comment by brandonkelly** [comment]
@kara-todd it’s working as expected (explanation: https://github.com/craftcms/cms/issues/6597#issuecomment-671638899).

**Comment by kara-todd** [comment]
@brandonkelly To clarify. I am indeed using `immediately: false` but this returns `https://www.domain.com/index.php?p=actions/assets/generate-transform&transformId=1` which in my case is a standalone gatsby site. Therefore, the image transforms fail to actually get called. I need `https://cms.domain.com/index.php?p=actions/assets/generate-transform&transformId=1` to be hit to actually generate the transform. On subsequent loads the graphql query will return the actual asset url. 

Sorry, I guess I'm not entirely following the explanation here. 😬 

If the transform url returns my public domain (thus never completing the transform  action CP request) how do the images get generated? My graphql calls were timing out without `immediately: false` due to some large transforms needing to be run.

**Comment by brandonkelly** [comment]
@kara-todd Thanks, and good point. Just implemented the original ask in #5553 (use the base CP URL when in headless mode) for the next Craft 3.6 release – https://github.com/craftcms/cms/issues/5553#issuecomment-762505392.
